### ISSUE TITLE
Jump Errors ?

### ISSUE THREAD

**Issue by HoratioGamer** [MeteorDevelopment/starscript#15]
So I found that StarScript can apparently execute scripts, and it had some interesting commands in it, so, I was digging deeper when I found this inside the very cute machine code virtual machine someone has created...

[starscript](https://github.com/MeteorDevelopment/starscript/tree/master)/[src](https://github.com/MeteorDevelopment/starscript/tree/master/src)/[main](https://github.com/MeteorDevelopment/starscript/tree/master/src/main)/[java](https://github.com/MeteorDevelopment/starscript/tree/master/src/main/java)/[org](https://github.com/MeteorDevelopment/starscript/tree/master/src/main/java/org)/[meteordev](https://github.com/MeteorDevelopment/starscript/tree/master/src/main/java/org/meteordev)/[starscript](https://github.com/MeteorDevelopment/starscript/tree/master/src/main/java/org/meteordev/starscript)

/Starscript.java

Lines 70-72

```
                case Jump:              { int jump = ((script.code[ip++] << 8) & 0xFF) | (script.code[ip++] & 0xFF); ip += jump; break; }
                case JumpIfTrue:        { int jump = ((script.code[ip++] << 8) & 0xFF) | (script.code[ip++] & 0xFF); if (peek().isTruthy()) ip += jump; break; }
                case JumpIfFalse:       { int jump = ((script.code[ip++] << 8) & 0xFF) | (script.code[ip++] & 0xFF); if (!peek().isTruthy()) ip += jump; break; }
```

I do not speak Java specifically, but from my knowledge of other computer languages, machine code and creating virtual machines, it looks like the repeated expression to compute "jump" is an error.   I think it should have been written instead as:

 `int jump = ((script.code[ip++] & 0XFF) << 8) | (script.code[ip++] & 0xFF);`

Which still makes the compiled script machine code syntax not usual in the sense that the instruction is coded  Jump  HighByte   LowByte, but whatever.   But at least with the correction it has both a functional high and a low byte.   As written:

`((script.code[ip++] << 8) & 0xFF)` = 0,  always, so there can never be a high byte of the jump address.

...unless JAVA is different from every other language I know.   

I suspect no one has ever created a program substantially over 256 bytes compiled, otherwise it would have failed on a Jump statement.  This might be an opportunity to change it to    Jump LowByte HighByte  in both the compiler and decompiler as well as the machine code virtual machine.   

I find this under-use situation very sad because it seems to me this is a really useful script language, far easier to use than JSMACROS.   With the addition of a few more instructions, which I believe is something I can do, even in JAVA, it could be really really cool and useful.   I have not seen the compiler yet, so, I do not know what the higher level language looks like, if there is one, but I would program this in Assembly: OpCodes and arguments instead of bytes.   I have written machine code, assembly, and coded virtual machine simulators.   I have decades of experience in this.   

Can someone please tell me how these scripts are compiled and what the source code looks like ?   

And can any of this be executed from Meteor ? 

 (That is what I was looking for in the code, found the execution section first... and what looks like an error repeated 3 times.)

**Comment by MineGame159** [comment]
Yeah looks like thats wrong, good catch. Strange how nobody ran into it before since I saw / made quite long starscript expressions, guess they werent 256 bytes long.

**Comment by HoratioGamer** [comment]
OK, I have not figured this out clearly, but, if this is just translating StarScript formatting expressions into this bytecode, then, my first wild guess would be, where would Jump be used in such a linear processing problem ?

Maybe Jump is just not used.   

Jump and JumpIfTrue were the two most interesting commands here, because it means I can program Algorithms in this language.   Algorithm = program a box in meteor, meaning, people with a wealth of programming experience can program a Minecraft Entity to do a great number of things without ever learning JAVA and all the sticky non-sequential, implied aspects of it.

**Comment by MineGame159** [comment]
The [syntax wiki](https://github.com/MeteorDevelopment/starscript/wiki) states that you can call predefined functions but you cant create new ones. And the jump instructions are for the conditional (aka ternary) operator. You can see some basic examples in the README of Starscript or the presets of text HUD in Meteor.

**Comment by HoratioGamer** [comment]
Which HUD text exactly ?   I have looked though them find them very elementary, giving me know clue how to do more interesting stuff.

**Comment by MineGame159** [comment]
I dont know from the top of my head but if you look through the different presets you will see some that use the conditional operator and call functions.

**Comment by HoratioGamer** [comment]
(aka ternary) operator you mean this?   <condition> ? <expression when condition is true> : <expression when condition is false>

Does not exist in any of the 20-languages I know.   Makes for very hard to read code, no room for proper comments.

**Comment by MineGame159** [comment]
Well I dont know which languages you know but it is present in MANY languages I know. Also Starscript doesnt even have comments. I am sure you have noticed it is just a text formatting language. You dont have multiple lines to work with, cant create variables, loops, functions or objects. So having a proper if statement is just not necessary and wouldnt even work in the context of Starscript.

**Comment by HoratioGamer** [comment]
Oh, but StarScript could be so much more than a formatting language, easily.   Yes, I am volunteering.
   
All it needs is a command escape character, and a command that reads source code script from a filename, like Baritone reads schematics.   The source code becomes script bytecode right after it is read, and existing StarScript code executes it.   What would be new is the potential to create a script that does not end, that is an algorithm like a poor-man's mod.  Attempting to execute the same script again, while it is still running should be a signal to kill the previous instance.   So if it is bound to a key in a macro, it automatically toggles on and off when the key is pressed.   
 
I just have to be clear that minecraft variables are read on the fly, not once before execution from which point they are static -- static for the execution of the script is not as versatile.   

How would it do anything?  By formatting strings that are Baritone commands.

A watcher (polling between instructions) in an executing instance of StarScript would have to monitor for messages to the log from Baritone that say "ok cancel" or whatever, so, the Baritone Cancel command can kill the executing StarScript scripts also.

**Comment by HoratioGamer** [comment]
OK, I see where the ternary operator is used in 4 HUD elements.

Here is an example of the code in the Assembly/Structured Language that would go with this bytecode:

```

// Comment
// Script that does x.
// 
Label Top     //  A new command that is a compiler directive, it does not translate into bytecode
                     //  creates a marker at this instruction pointer location that a JUMP can use later
                     //  This label is not used in this program
                     //   Lots of extra labels are created automatically and resolved to instruction pointer addresses when the
                     //   first pass of compilation is done.   
                     //   Second pass completes the authoring of bytecode.
//
// Infinite loop that keeps trying to get the player to go to either x =0 or x= 50000, whichever is closer
//
// This is just one small element of a program for a bot that might be programmed to return home 
//  when it completes a mission.  
//  This code is extremely simplified.
//
//  To help explain how structured language blocks like If-Else-EndIf blocks or While-EndWhile are coded into 
//  existing instructions during compilation, comments have been added.
//
While {True}                // codes to 
                                   //     Label SS-While-Top-0001   (because it is the first structure block the compiler found)
                                   //     True
                                   //     JumpIfFalse SS-While-Exit-0001

   LoadLive baritone.process_name

   //
   // If Baritone is not running, begin composing a new Baritone command
   If { baritone.process_name ==  ""}                  // codes to 
                                   //     LoadLive baritone.process_name
                                   //     Constant ""
                                   //     Equals
                                   //     JumpIfFalse SS-If-ElseIf-0002
        Constant "goto " 

        If { player.pos.x ==   0 }          // Codes to
                                                       //     LoadLive  player.pos.x     A new command to load a live string into the stack, not a static one
                                                       //     Constant 0
                                                       //     Equals
                                                       //     JumpIfFalse  SS-If-ElseIf-0003
            Jump Arrived
        ElseIf {player.pos.x ==   50000  }     // Codes to
                                                       //     Jump SS-If-EndIf-0003
                                                       //     Label SS-If-ElseIf-0003
                                                       //     LoadLive  player.pos.x     A new command to load a live string into the stack, not a static one
                                                       //     Constant 50000
                                                       //     Equals
                                                       //     JumpIfFalse  SS-If-ElseIf-0004
            Jump Arrived
        ElseIf  {player.pos.x  <  25000  }                        // codes to 
                                                //      Jump SS-If-EndIf-0003 
                                                //      Label SS-If-ElseIf-0004
                                                //     Constant 25000
                                                //     Equals
                                                //     JumpIfFalse  SS-If-ElseIf-0005
           Constant 0
        Else                                // codes to 
                                                //      Jump SS-If-EndIf-0003 
                                                //      Label SS-If-ElseIf-0005
           Constant 50000
        EndIf                                       // codes to 
                                                       //      Label SS-If-EndIf-0003 
        Append 
        Constant " ~ ~ "
        Append
        Baritone                     // codes as:
                                          // Constant "#"             This assures nothing can go to game chat by accident.   
                                          //                                           A totally different Chat command might be used for that.   
                                          // Exch                         Exchange top two stack elements
                                          // Append
                                          // pop top string and send it to Baritone
        
    EndIf                              // codes as 
                                          //       Label SS-If-ElseIf-0002    because this if had no else
                                          //       Label SS-If-EndIf-0002   so that if there had been an else if would give the true block 
EndWhile                    // codes to 
                                   //     Jump SS-While-Top-0001
                                   //     Label SS-While-Exit-0001  
Label Arrived

End
```

New commands might be added to the list of Reverse Polish Notation / stack using commands (like high end HP calculators from the 1980's or PostScript from 1980s-present) commands.   

Dupe or Push        -- duplicate the top item on the stack so the top two items are the same as the top item before.
Exch                      -- exchange the top two items.

I have years of experience writing code that wrote PostScript code, those commands were essential.


More new structured language blocks can be made (/ means the other part of the block is on a new line that encloses a block of code:

     `Repeat / Until {}`    or     `Do / While {}`      or both
     `Loop / EndLoop`                        -- Equivalent to While {true}
     `For / Next` or `For / EndFor`               -- counted loops with a loop variable stored in a separate storage.

     `ForEach / EndForEach`                -- awfully useful in TCL to parse lists of arguments.   

     `Break n`                                    -- break out of n layers of loop/while/repeat/for structures
     `Continue n`                               -- the single Break/Continue can lead to the need for Jump or carrying temp variables.  
     `Switch {} / Case / EndCase / EndSwitch`

These are just more automatically-generated condition tests and jumps to automatically generated labels.  Trivial stuff. 

True User functions Function / Return / EndFunction can be created.  There are two conditions:   The function must be defined before it is used to keep the compilation to 2 passes.   The user is responsible for maintaining the stack integrity, as in, getting the right number of arguments into the stack before the call, pulling the right number of arguments from the stack in the call, pushing the right data onto the stack after, but not the push and pop of return addresses.  If the Function Returns one value, it is coded as an Exch / Jump , pulling the return address from the stack.

This is another PostScript thing, substitution macros:

```
/Name  
   substitution code 
Def
```

So in future, one just has to say 

`Name `

and the compiler knows to insert the expansion.    Again, substitution macros must be defined before they are used so that compilation is 2 passes.  

The compiler needs a stack that counts how many nested structure blocks we are inside, to check that any block-ending keyword matches the current top of the structure stack -- detect syntax errors,  generate the automatic labels,  and also to implement Break n and Continue n.    

When running, it would probably be most convenient if StarScript had an array of structures with Data, NameContext, StructureLevel.    So that local variables, like for For loops or inside Functions could be unique to their context and structure nesting level.   Once out of the function or block it was valid in, it becomes invalid.

**Comment by HoratioGamer** [comment]
StarScript has the ability to run programs that compile to up to 65k of bytecode -- that is a HUGE program.    Controllers for entire buildings ran on less program space for a long time.   The Space Shuttles each had 5 main computers with 12k of program space each... only in aggregate are they more program space than StarScript.    

Trust me, a really competent bot could be programmed.  Meteor modules could be activated and deactivated.

**Comment by MineGame159** [comment]
I am not entirely against a scripting solution for Meteor but Starscript is not it. This is a formatting language, nothing more.

**Comment by HoratioGamer** [comment]
The problem is, the JAVAheads in the Meteor discord cant imagine anything that is not JAVA being useful.   

The most useful language ever was the first Procomm scripting language, which Aspect might be a superset of, never used Aspect, I had a better internet by then.   That language was designed to negotiate (talk and listen) telephone modem connections to different services, however, they accidentally put enough commands in it to make it more widely useful.  It could conduct entire sessions, automatically transferring data, files, updating off-line email servers, navigate remote access interfaces that required selections or commands before an IP/PPP connection could be made.   There was no length limit I was aware of on those scripts.   I had some very long and useful ones.  

What I suggest is adding a few things no one is obligated to use, an optional plus.   Any expansion of StarScript by the plus would be completely transparent to its current uses.   One of those would be an unused script compiler function/procedure that takes the code above and turns it into bytecode -- a simple two-pass linear process I have coded at least 3 times before, including the down-authoring of structured language blocks into available commands like Label and Jump.   

Yes, it may ultimately require another interface in Meteor, they can call it Meteor Script for all I care, or a command in Meteor .commands to access what I have suggested.   We can cross that bridge when we get to it after there is a powerful ability available.

The Meteor developers may not have realized that they have accepted a non-JAVA bytecode into the project, may have no idea how powerful it could be with an Assembly Language compiler in addition to the string formatting compiler that must exist somewhere, though I still have not found it.   

The advantage of the bytecode you have developed is, it has no implied execution, it is all on the surface, step by step, and all execution is explicitly sequential.  I have experienced teaching a JAVAhead assembly language.   I understand it is unfamiliar, but, "it is linear, in small steps, just take it step by step" works.   Understanding it one statement at a time, it forms the words in the story that is the process, what it does.  Yes low level languages have a forest for the trees problem, but it is better than Java's Forest.forest.smallforest problem where you have to read 30-100 lines of code in 100 files to find the flippin trees of what is it is actually doing right now.   There is no other library of inherited functions in your language, no parent class to go read about, which might have another parent class.  There is no abstraction.  There is nothing happening in the background automatically once one understands a stack.   It reads like reverse Polish notation HTML, like Turtle Logo Graphics, like PostScript, like dozens of simple but powerful scripting languages that have existed in the past.   

For those who have bent their mind around JAVA's inherent twist and concealment, they cannot understand how a simple linear language is far more understandable by more people, programmable by anyone with any programming background.  

Meteor devs will never allow a non-Java scripting language written from scratch.   They will just dismissively say, go read JSMACROS, which they have, and I have, and I quickly determined it was not accessible to me, and not the most accessible solution.

They will never accept anything from me.  I lack the general ability to write it from scratch in JAVA.  Starscript is an existing platform where many parts of it are far more linearly written (I like the writing style), that is why I can read them and am confident I can add to them.

Think of this from the point of view of a player.   They are an ant blissfully unaware of the layered Gordian Knot of JAVA that is the game and all the mods.   They play the game and accomplish goals.  They flip switches (modules) on and off.    This scripting language could be far closer to the world view of the surface ant.

**Comment by HoratioGamer** [comment]
Step one is to add limits checking to current commands of the current StarScript execution "run" command, so it can raise a StarScript execution error.   Yes, the compiler probably has these, but, two layers of error trapping never hurt anything.

Error trapping is needed if people are to write things that become bytecode.   Bottom up, no errors.   No program written in Forth ever duplicated a nuclear power plant.

**Comment by MineGame159** [comment]
Fixed in 77fa442a8cd2499914dec0bbe454a4ad4da85dc5.

**Comment by HoratioGamer** [comment]
Yeah, I misread the code at first.    Jumps are relative, add an offset, not absolute instruction addresses.

Jumps are only used in () ? () : () ternary statements.   The second or 3rd () would have to code to more than 256 bytes to see the error.


Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by GreemDev** [comment]
> The problem is, the JAVAheads in the Meteor discord cant imagine anything that is not JAVA being useful.

No actually minegame said he wouldn't be opposed to scripting in Meteor, it's just that Starscript **isn't it**.

> The Meteor developers may not have realized that they have accepted a non-JAVA bytecode into the project, may have no idea how powerful it could be with an Assembly Language compiler in addition to the string formatting compiler that must exist somewhere, though I still have not found it.

who the hell cares

> For those who have bent their mind around JAVA's inherent twist and concealment, they cannot understand how a simple linear language is far more understandable by more people, programmable by anyone with any programming background.

We get it, you don't like Java, how is this relevant

> Meteor devs will never allow a non-Java scripting language written from scratch. They will just dismissively say, go read JSMACROS, which they have, and I have, and I quickly determined it was not accessible to me, and not the most accessible solution.

[Can you read? Like, at all?](https://github.com/MeteorDevelopment/starscript/issues/15#issuecomment-1711510292)
and before you say "But it'd probably be written in Java", who cares? this is a project primarily used for a block game (written in java) client mod. there's absolutely no need to write it at a low level to then reference from java code.

> They will never accept anything from me. I lack the general ability to write it from scratch in JAVA. Starscript is an existing platform where many parts of it are far more linearly written (I like the writing style), that is why I can read them and am confident I can add to them.

Ok, then don't contribute


By the way, it's `Java`, or even `java`. It's not an acronym.

This entire conversation reads like someone who is really desperate to let people who use Java know that they hate Java.
Shut up and go back to writing Assembly in vim.

**Comment by HoratioGamer** [comment]
>>No actually minegame said he wouldn't be opposed to scripting in Meteor, it's just that Starscript isn't it.

Which means writing from scratch in Java.  Its unlikely I could do that.

>>We get it, you don't like Java, how is this relevant

Java is the first language I have come across that I could not learn by reading it, atleast not from mods.  It has been 4 years on and off.   Most other languages took a couple days to have a solid start, developing code, learning by debugging from there.   Part of the problem for me is, minecraft+mods is a really big collection of code to start into.   And all the code is someone else's baby.   I put errors in my own code, it us just an inconvenience.   I am reluctant to mess up someone else's code.

I am impressed by how few lines the script compiler is, between the Lexer, Parser, and Compiler and maybe a couple more modules I have hopped back and forth between in trying to figure out how it does what it does.   Maybe someone took a compilers course in uni, I never did.   The design solution seems not-obvious.  It is the first time I am not seeing it as being obviously easier to program an algorithm in another language.

It is relevant because I would like to program in Java to solve block game problems.   The block game problem I would like to solve is something that makes it very easy for others to program to solve block game problems so many others could do it.   I seek to open doors.   If driving a car were as hard as getting into coding with minecraft, yes, everyone here would drive, but few would as a fraction of the population.

I would like there to be more programmers in the world by lowering the bar to entry.   I have taught more than 1,000 people how to program.   I would like to get another 1,000 into programming before I am done.   This game is an avenue.   Their motivation is high, if the bar is low enough, many will become programmers.   And they will come to Meteor to do it.   Meteor will be their Procomm, their Expect, their Turtle Logo Graphics, but in a 3d multiplayer game.

I bitch about Java partially because of the programming style used in mods.  Less than 5% of lines are comments.  For the most part, the code is written in riddles.   This is not new, seen this style for decades, just never met a language that made it quite so easy to write riddles before.   I have been gifted books on Java and it reads as an entirely different language there.   Programmers who resist writing full documentation for users is not new either.   It is not helping others get into coding or using.

What I have learned in Meteor, I have passed on to apparently smart people in other groups.   So far everyone has thanked me for explaining it.

>>there's absolutely no need to write it at a low level to then reference from java code.

To lower the bar, so programmers of the script do not have to read the github to meet their block game goals.    The script I wrote above was derived from the StarScript instruction set that looked like machine code to me.   If i were to start from scratch, I would look forward not back.  I spent an evening learning and teaching another programmer, a Java wiz, RISC-V a few years ago.   Or maybe I will look at what is taught in grade schools now and let that influence the style/syntax.

If I can get anarchy players to code their first program, because they want to, watch it run in 3d,  feel success, that is the goal.

>>This entire conversation reads like someone who is really desperate to let people who use Java know that they hate Java.

I hate that Java in mods is such a high bar just to understand what it is doing.    I would like those who find it easy to appreciate, it really is not easy.   Really.

You have a moment now, before Lambda is rewritten to capture a lot more users.  Make it easy for users, let the other options be under-documented and their community insulting.

Make it a priority that Meteor be the easy gateway to programming, by saying just once, ok, write it, we will help you get the Java correct.

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________
